File: python/src/skillflag/frontmatter.py
"""Minimal SKILL.md frontmatter parser (port of typescript/src/shared/frontmatter.ts).

This is deliberately NOT full YAML: the block must start with ``---`` on the
first line and end at the next ``---`` line; each line is split at the first
``:``; one pair of matching surrounding quotes is stripped from the value.
"""

from __future__ import annotations

import re

_FRONTMATTER_RE = re.compile(r"^---\s*\r?\n(.*?)\r?\n---(?:\r?\n|$)", re.DOTALL)
_LINE_SPLIT_RE = re.compile(r"\r?\n")


def _strip_yaml_quotes(value: str) -> str:
    if (value.startswith('"') and value.endswith('"')) or (
        value.startswith("'") and value.endswith("'")
    ):
        return value[1:-1].strip()
    return value


def parse_frontmatter(content: str) -> dict[str, str]:
    match = _FRONTMATTER_RE.match(content)
    if match is None:
        return {}

    fields: dict[str, str] = {}
    for line in _LINE_SPLIT_RE.split(match.group(1)):
        if not line.strip():
            continue
        idx = line.find(":")
        if idx == -1:
            continue
        key = line[:idx].strip()
        value = _strip_yaml_quotes(line[idx + 1 :].strip())
        if key and value:
            fields[key] = value
    return fields

File: go/internal/core/frontmatter.go
package core

import (
	"regexp"
	"strings"
)

// frontmatterRe mirrors the reference implementation's regex:
// /^---\s*\r?\n([\s\S]*?)\r?\n---(?:\r?\n|$)/
var frontmatterRe = regexp.MustCompile(`^---\s*\r?` + "\n" + `([\s\S]*?)\r?` + "\n" + `---(?:\r?` + "\n" + `|$)`)

var lineSplitRe = regexp.MustCompile(`\r?\n`)

func stripYamlQuotes(value string) string {
	if (strings.HasPrefix(value, `"`) && strings.HasSuffix(value, `"`)) ||
		(strings.HasPrefix(value, "'") && strings.HasSuffix(value, "'")) {
		if len(value) < 2 {
			// A lone quote character: JS "\"".slice(1, -1) yields "".
			return ""
		}
		return strings.TrimSpace(value[1 : len(value)-1])
	}
	return value
}

// ParseFrontmatter extracts the minimal key/value frontmatter block from a
// SKILL.md document. This deliberately is NOT a full YAML parser; it ports
// typescript/src/shared/frontmatter.ts exactly.
func ParseFrontmatter(content string) map[string]string {
	fields := map[string]string{}

	match := frontmatterRe.FindStringSubmatch(content)
	if match == nil {
		return fields
	}

	for _, line := range lineSplitRe.Split(match[1], -1) {
		if strings.TrimSpace(line) == "" {
			continue
		}
		idx := strings.Index(line, ":")
		if idx == -1 {
			continue
		}
		key := strings.TrimSpace(line[:idx])
		value := stripYamlQuotes(strings.TrimSpace(line[idx+1:]))
		if key != "" && value != "" {
			fields[key] = value
		}
	}

	return fields
}
